### DISCUSSION TITLE
Building a Rust Crate

### DISCUSSION THREAD

**Discussion by schroe-ja** [openwpm/OpenWPM#1072] [Ideas]
I found OpenWPM and want to use it for our project. Is it possible to implement a Rust crate? Or an API that is not bound on a specific programming language?

**Comment by vringar** [comment]
Hey @EvilWatermelon,

unfortunately OpenWPM relies heavily on Python as we expose most of our functionality by allowing users to extend the `BaseCommand` class and then adding their class to the `CommandSequence`.

If you want to use OpenWPM from Rust, please have a look at [Py03](https://pyo3.rs/v0.20.0/python_from_rust), but please keep in mind that there might be some weirdness related to the fact that OpenWPM is a multi-*process* application, so some Python code would likely be executed outside of the the Rust wrapper.

**Comment by schroe-ja** [comment]
@vringar ok, thanks for your fast answer. I think using mqtt should be easier while using OpenWPM as an external service.
